File: h10/h10.rb
Pair = Struct.new(:fst,:snd)

def summingPairs(xs, sum)
  the_pairs1 = []
  the_pairs2 = []
  the_pairs = []
  len = xs.length

  reader, writer = IO.pipe 
  
    fork do
     for i in 0..(len-1)
      for j in (i+1)..(len/2)
        if xs[i] + xs[j] <= sum
            writer.puts(Pair.new(xs[i],xs[j]))
        end
      end
    end
  end
  
    fork do
      for i in 0..(len-1)
         for j in (len/2)..(len-1)
             if xs[i] + xs[j] <= sum
                  writer.puts(Pair.new(xs[i],xs[j]))
             end
          end
        end
      end
  Process.waitall
  writer.close
  while message = reader.gets
    the_pairs.push(message)
  end

  return the_pairs
end
require 'date'
require_relative 'collection'



puts (summingPairs(INPUT,2020))
# puts "Starting at:   #{DateTime.now.sec} seconds, #{DateTime.now.strftime("%9N")} nanoseconds"
puts (summingPairs(INPUT,2020).include?(Pair.new(626, 1030)))
# puts "Ending at:     #{DateTime.now.sec} seconds, #{DateTime.now.strftime("%9N")} nanoseconds"

File: h10/h10.clj
(load-file "collection.clj") 
(defn summingPairs [xs sum]
(print("dfgdfda"))
  (letfn [(summingPairsHelper [xs the_pairs]
            ;; If `xs` is empty, we're done.
            
            (if (empty? xs) the_pairs
            
                (let [[fst & rest] xs]
                
                   (recur
                   rest
                   (doall 
                    (concat the_pairs
                            (for [snd rest ;; For each `snd` in `rest`...
                                  :when (<= (+ fst snd) sum)]
                              [fst snd]))))))
           (let [first_half  (future (summingPairs (subvec xs 0 (/ (count xs) 2)) sum ))
                  second_half (future (summingPairs (subvec xs (/ (count xs) 2)) sum ))]
            (concat first_half second_half)       
            )                   
                              
                              
                              
                             
                         
  )]
    (summingPairsHelper xs [])))


    (defn middle-value [vect]
  (when-not (empty? vect)
    (vect (quot (count vect) 2))))
(print(summingPairs input 2020))
    ; (print(middle-value input))
; (print(subvec input 0 (/ (count input) 500) ))

(defn summingPairs [xs sum]
  (letfn [(summingPairsHelper [xs the_pairs]
            ;; If `xs` is empty, we're done.
            (if (empty? xs) the_pairs
                ;; Otherwise, decompose `xs` into the `fst` element
                ;; and the `rest`.
                (let [[fst & rest] xs]
                  ;; We use the `recur` form to make the recursive call.
                  ;; This ensures tail call optimisation
                  (recur
                   rest
                   ;; Concatenate `the_pairs` we have so far with the sequence
                   ;; of every `[fst snd]` where `snd` is in `rest` with
                   ;; `fst + snd <= sum`. The `doall` outside the `concat`
                   ;; forces it to be calculated immediately; without this,
                   ;; we get a (lazy) buildup of `concat`'s which may
                   ;; cause a stack overflow when looking at the result.
                   (doall 
                    (concat the_pairs
                            (for [snd rest ;; For each `snd` in `rest`...
                                  :when (<= (+ fst snd) sum)]
                              ;;... put `[fst snd]` into this sequence.
                              [fst snd])))))))]
    (summingPairsHelper xs [])))
(println (str
          "Starting at:   "
          (.getSecond (java.time.LocalDateTime/now))
          " seconds, "
          (.getNano (java.time.LocalDateTime/now))
          " nanoseconds"))
; (println (summingPairs input 2020))
(println (str
          "Ending at:     "
          (.getSecond (java.time.LocalDateTime/now))
          " seconds, "
          (.getNano (java.time.LocalDateTime/now))
          " nanoseconds"))